File: source/ClubResult.m
//
//  ClubResult.m
//  Tournament
//
//  Created by Paul Trunz on 18.04.15.
//  Copyright 2015 __MyCompanyName__. All rights reserved.
//

#import "ClubResult.h"
#import "PlayerResult.h"
#import "SmallTextController.h"


@implementation ClubResult

- (ClubResult *) initWithName:(NSString *)aName;
{
	self = [super init];
	name = aName;
	results = [NSMutableArray array];
	
	return self;
}

+(ClubResult *) with:(NSString *)aName;
{
	ClubResult* result = [[ClubResult alloc] initWithName:aName];
	
	return result;
}

-(void)add:(SinglePlayer *)player series:(Series *)ser points:(long)points;
{
	NSString * serPlayer = [NSString stringWithFormat:@"%@: %@", [ser seriesName],
													[player longName]];
	PlayerResult *singleResult = [PlayerResult for:serPlayer points:points];
	total = total+points;
	[results addObject:singleResult];
}

- (NSComparisonResult)compare:(ClubResult *)otherClub {
	long otherTotal = [otherClub total];
	if (total < otherTotal) {
		return NSOrderedDescending;
	} else if (total > otherTotal) {
		return NSOrderedAscending;
	} else {
		return NSOrderedSame;
	}
}

- (long) total;
{
	return total;
}

- (void)appendAsLineTo:(id)text;
{
	[text appendText:[NSString stringWithFormat:@"%@\t%ld\n", name, total]];
}

- (NSArray *) results;
{
	return results;
}

- (NSString *) name;
{
   return name;
}
@end

File: ClubResult.swift
//
//  ClubResult.swift
//  Tournament
//
//  Created by Paul Trunz on 17.11.18.
//

import Foundation

struct PlayerResult {
   let playerInSeries: String
   let points: Double
   init(for plInSer: String, points: Double) {
      playerInSeries = plInSer
      self.points = points
   }
   
   func appendAsLine(to text:SmallTextController) {
      text.appendText("\t\t\(playerInSeries)\t\(points)\n")
   }
}

class ClubResult : Comparable {
   static func <(lhs: ClubResult, rhs: ClubResult) -> Bool {
      return lhs.total < rhs.total
   }
   
   static func ==(lhs: ClubResult, rhs: ClubResult) -> Bool {
      return lhs.total == rhs.total
   }
   
   let name:String
   var results = [PlayerResult]()
   private(set) var total = 0.0
   
   init(for name:String) {
      self.name=name
   }
   
   func add(_ player:SinglePlayer, series:Series, points:Double) {
      let seriesPlayer = "\(series.seriesName()!): \(player.longName()!)"
      let plRes = PlayerResult(for:seriesPlayer, points:points)
      // TODO: make it a real double
      results.append(plRes)
      
      total += points
   }
   
   func appendAsLine(with rank:String, to text:SmallTextController) {
      text.appendText("\t\(rank)\t\(name)\t\(total)\n")
   }
   
   func appendDetails(to text:SmallTextController) {
      for result in results {
         result.appendAsLine(to:text)
      }
   }
}

